windows/RNFetchBlob/RNFetchBlob.cpp

```diff
@@ -0,0 +1,1637 @@
+//
+void RNFetchBlob::ConstantsViaConstantsProvider(winrt::Microsoft::ReactNative::ReactConstantProvider& constants) noexcept
+{
+	// RNFetchBlob.DocumentDir
+	constants.Add(L"DocumentDir", to_string(ApplicationData::Current().LocalFolder().Path()));
+
+	// RNFetchBlob.CacheDir
+	constants.Add(L"CacheDir", to_string(ApplicationData::Current().LocalCacheFolder().Path()));
+
+	// RNFetchBlob.PictureDir
+	constants.Add(L"PictureDir", UserDataPaths::GetDefault().Pictures());
+
+	// RNFetchBlob.MusicDir
+	constants.Add(L"MusicDir", UserDataPaths::GetDefault().Music());
+
+	// RNFetchBlob.MovieDir
+	constants.Add(L"MovieDir", UserDataPaths::GetDefault().Videos());
+
+	// RNFetchBlob.DownloadDirectoryPath - IMPLEMENT for convenience? (absent in iOS and deprecated in Android)
+	constants.Add(L"DownloadDir", UserDataPaths::GetDefault().Downloads());
+
+	// RNFetchBlob.MainBundleDir
+	constants.Add(L"MainBundleDir", to_string(Package::Current().InstalledLocation().Path()));
+}
+
+// createFile
+winrt::fire_and_forget RNFetchBlob::createFile(
+	std::string path,
+	std::wstring content,
+	std::string encoding,
+	winrt::Microsoft::ReactNative::ReactPromise<std::string> promise) noexcept
+try
+{
+	bool shouldExit{ false };
+	Streams::IBuffer buffer;
+	if (encoding.compare("uri") == 0)
+	{
+		try
+		{
+			winrt::hstring srcDirectoryPath, srcFileName;
+			splitPath(content, srcDirectoryPath, srcFileName);
+			StorageFolder srcFolder{ co_await StorageFolder::GetFolderFromPathAsync(srcDirectoryPath) };
+			StorageFile srcFile{ co_await srcFolder.GetFileAsync(srcFileName) };
+			buffer = co_await FileIO::ReadBufferAsync(srcFile);
+		}
+		catch (...)
+		{
+			shouldExit = true;
+		}
+	}
+	else if (encoding.compare("utf8") == 0)
+	{
+		buffer = CryptographicBuffer::ConvertStringToBinary(content, BinaryStringEncoding::Utf8);
+	}
+	else if (encoding.compare("base64") == 0)
+	{
+		buffer = CryptographicBuffer::DecodeFromBase64String(content);
+	}
+	else
+	{
+		promise.Reject("Invalid encoding");
+		shouldExit = true;
+	}
+
+	if (!shouldExit)
+	{
+
+		winrt::hstring destDirectoryPath, destFileName;
+		splitPath(path, destDirectoryPath, destFileName);
+
+		auto folder{ co_await StorageFolder::GetFolderFromPathAsync(destDirectoryPath) };
+
+		try
+		{
+			auto file{ co_await folder.CreateFileAsync(destFileName, CreationCollisionOption::FailIfExists) };
+			auto stream{ co_await file.OpenAsync(FileAccessMode::ReadWrite) };
+			co_await stream.WriteAsync(buffer);
+		}
+		catch (...)
+		{
+			promise.Reject("EEXIST: File already exists."); // TODO: Include filepath
+			shouldExit = true;
+		}
+	}
+	if (!shouldExit)
+	{
+		promise.Resolve(path);
+	}
+	co_return;
+}
+catch (const hresult_error& ex)
+{
+	hresult result{ ex.code() };
+	if (result == 0x80070002) // FileNotFoundException
+	{
+		promise.Reject("ENOENT: File does not exist and could not be created"); // TODO: Include filepath
+	}
+	else
+	{
+		promise.Reject("EUNSPECIFIED: Failed to write to file.");
```

Comment:
probably want to pass the hresult at least. You can use the `hresult_error`'s `message()` member function to get a string representation of the error